Seminar3

KWY0218 opened the pull request:
- [x] 필수 과제
- [x] 성장 과제
- [x] 도전 과제

[Read Me](https://github.com/SAMSUNG-ZZANG/Galaxy-Tab-Wonyong/blob/seminar3/README.md)

seminar 4주차에 도전과제 완

KWY0218 committed:
[ADD] Add text style

textview 공통 color, font 설정

KWY0218 committed:
[CHORE] Edit theme name

customTextView -> custom_text 변경

KWY0218 committed:
[Add] Add multiple button

버튼 select 기능 추가

KWY0218 committed:
[Add] Show circle image with glide

BindingAdapter, Glide 이용해서 이미지 원형으로 출력

KWY0218 committed:
[ADD] Add bottom navigation bar view pager

KWY0218 committed:
[ADD] Add home fragment

tabLayout + viewPager 추가

KWY0218 committed:
[CHORE] Edit tab layout indicator color

텝 레이아웃 인디케이터 컬러 수정

KWY0218 committed:
[REFACTOR] Refactor nested scroll problem

중첩 스크롤 문제 해결

KWY0218 committed:
Update README.md

KWY0218 committed:
Update README.md

mdb1217 commented:
도전과제 말하는거겠쥥...ㅋㅋㅋㅋ??? 암튼 굿굿

mdb1217 commented on app/build.gradle:
```diff
@@ -38,6 +38,10 @@ android {
 dependencies {
+    // Glide
+    implementation 'com.github.bumptech.glide:glide:4.13.0'
+    annotationProcessor 'com.github.bumptech.glide:compiler:4.13.0'
```
`annotationProcessor`의 경우 자바식 어노테이션 프로세서입니당..!

```suggestion
    kapt 'com.github.bumptech.glide:compiler:4.13.0'
```

로 바꿔주시면 좋을 것 같아요 :)

KWY0218 replied:
오오... 처음 알았네요..!
근데 coli로 수정해서 삭제했습니다.. 😅

mdb1217 commented on app/src/main/java/com/example/sopt_seminar/ui/HomeFragment.kt:
```diff
@@ -3,21 +3,34 @@ package com.example.sopt_seminar.ui
+    private fun initTabLayout() {
+        val tabList = listOf("팔로잉", "팔로워")
+
+        TabLayoutMediator(binding.homeTl, binding.homeVp) { tab, postion ->
+            tab.text = tabList[postion]
+        }.attach()
```
strings.xml에 있는 `<string-array>` 활용하면 좀 더 깔끔하게 표현가능할 것 같네용!

KWY0218 replied:
[dc389fe](https://github.com/SAMSUNG-ZZANG/Galaxy-Tab-Wonyong/pull/4/commits/dc389fe080ef574c2f637727da61f8058a7d5116)  수정했습니다!!

mdb1217 commented on app/src/main/java/com/example/sopt_seminar/ui/ProfileFragment.kt:
```diff
@@ -0,0 +1,45 @@
+    private val repoFragment = RepoFragment()
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        childFragmentManager.beginTransaction().replace(R.id.home_fragment_cv, followerFragment)
+            .commit()
```
fragment-ktx 이용해서 아래처럼 보다 프래그먼트 클래스만 참조해서 보다 간결하게 표현이 가능합니다..!
```suggestion
childFragmentManager.commit {
    replace<FollowerFragment>(R.id.home_fragment_cv)
}
```

mdb1217 replied:
이렇게 하는게 프래그먼트 클래스 객체를 직접 만들어서 활용하는 것에 비해 메모리 세이프하기도 하구요..!

KWY0218 replied:
[94ce90f](https://github.com/SAMSUNG-ZZANG/Galaxy-Tab-Wonyong/pull/4/commits/94ce90fb48c5e946e740388ba9e490eb2f613c06) 수정했습니다!!

mdb1217 commented on app/src/main/java/com/example/sopt_seminar/ui/ProfileFragment.kt:
```diff
@@ -0,0 +1,45 @@
+        }
+    }
+
+    val changeState = fun(view: View) {
+        val followerBtn = binding.followerBtn
+        val repoBtn = binding.repoBtn
+        when (view.id) {
+            R.id.follower_btn -> {
+                if (!followerBtn.isSelected) {
+                    followerBtn.isSelected = true
+                    repoBtn.isSelected = false
+                }
+            }
+            R.id.repo_btn -> {
+                if (!repoBtn.isSelected) {
+                    repoBtn.isSelected = true
+                    followerBtn.isSelected = false
+                }
+            }
+        }
+    }
```
변수를 함수화 시킬때 이것 또한 람다형으로 간결하게 표현 가능할듯
```suggestion
    val changeState : ((View) -> Unit) =  { view ->
        val followerBtn = binding.followerBtn
        val repoBtn = binding.repoBtn
        when (view.id) {
            R.id.follower_btn -> {
                if (!followerBtn.isSelected) {
                    followerBtn.isSelected = true
                    repoBtn.isSelected = false
                }
            }
            R.id.repo_btn -> {
                if (!repoBtn.isSelected) {
                    repoBtn.isSelected = true
                    followerBtn.isSelected = false
                }
            }
        }
    }
```

KWY0218 replied:
해당 부분은 없어졌지만 다음에 참고하겠습니다!!!!

mdb1217 commented on app/src/main/res/navigation/nav_graph.xml:
```diff
@@ -80,4 +61,22 @@
     </fragment>
+
+    <fragment
+        android:id="@+id/mainFragment"
```
보통 xml의 네이밍 컨벤션은 스네이크 케이스를 채택하는 경우가 많아요! 참고바람..!

KWY0218 replied:
[d31c9ed](https://github.com/SAMSUNG-ZZANG/Galaxy-Tab-Wonyong/pull/4/commits/d31c9edfe90539ad101d9795930eff7e9c9125f4) 아!!!!!! 실수했네요...! 수정했습니다!!

mdb1217 commented on app/src/main/java/com/example/sopt_seminar/util/BindingAdapter.kt:
```diff
@@ -14,13 +17,30 @@ import com.example.sopt_seminar.ui.adapter.RepoAdapter
     @JvmStatic
-    @BindingAdapter(value = ["fragmentManger", "childFragment"])
-    fun changeFragment(button: Button, fragmentManger: FragmentManager, childFragment: Fragment) {
+    @BindingAdapter(value = ["fragmentManger", "childFragment", "changeSelectState"])
+    fun changeFragment(
+        button: Button,
+        fragmentManger: FragmentManager,
+        childFragment: Fragment,
+        changeSelectState: (View) -> Unit
+    ) {
```
음 이건 개인적인 제 의견인데.. 너무 많은 작업을 바인딩 어댑터에 할당하고 있는게 아닌가 하는 느낌은 있습니다..😢😢 바인딩 어댑터의 경우 @JvmStatic으로 선언되서 메모리 상에 계속 남아있게 되어서.. 이런식으로 특정 세부기능을 정의하기보다는.. 좀 더 전역적으로 많이 활용되는 친구들이 바인딩 어댑터로 선언됩니당..😢 이런 부분은 그냥 뷰 코드 단에서 정의하는 게 좋지않을까 생각은 드네요..!

KWY0218 replied:
이번에 처음 써봐서 한번 다 바인딩 해보고 싶어서 해봤는데 막상 해보니까 번거롭기만 하고, 메모리 상에 계속 남아 있다는 생각은 하지 못했네요..! 감사합니다!!!

mdb1217 commented on app/src/main/java/com/example/sopt_seminar/util/BindingAdapter.kt:
```diff
@@ -14,13 +17,30 @@ import com.example.sopt_seminar.ui.adapter.RepoAdapter
+    fun showImg(imageView: ImageView, drawable: Int) {
+        Glide.with(imageView.context)
+            .load(drawable)
+            .override(100, 100)
```
코드단에서 사용되는 기본 수치단위는 px입니당.. dp 수치로 programmatically하게 주고싶다면, 아래 방식 참고해주세요!

- **dp To Pixel : (수치 값 * resources.displayMetrics.density).toInt()**

KWY0218 replied:
코드 단 기본 수치 단위는 px이였군요....! 감사합니당..!

mdb1217 commented on app/src/main/res/layout/repo_frame.xml:
```diff
@@ -25,23 +25,23 @@
 
             <TextView
                 android:id="@+id/repo_des_tv"
+                style="@style/custom_text"
```
보통 TextView나 EditText와 같은 뷰에서 사용되는 텍스트 속성들을 설정하는 스타일들은 `style` 속성이아닌 `TextAppearance` 속성에 많이 넣어줍니다..! style이 좀 더 많은 부분을 관장하는 속성이에요! 아래 링크 보시고 style과 TextApperance의 차이점을 한 번 알아보시길 바랍니다 :)

https://yoojh9.tistory.com/entry/10-1-Styles-and-Theme

KWY0218 replied:
[6c0c87a](https://github.com/SAMSUNG-ZZANG/Galaxy-Tab-Wonyong/pull/4/commits/6c0c87a172fe6f6280ec4457e3849a2bdcc457f0), [4167a84](https://github.com/SAMSUNG-ZZANG/Galaxy-Tab-Wonyong/pull/4/commits/4167a84c3293cd875fcd245737c46544e01805b0) 
수정했습니다..! 허허.. 리소스 관련 공부 좀 해야겠네요..

KWY0218 committed:
[CHORE] Change annotationProcessor to kapt

annotationProcessor를 kapt로 변경

KWY0218 committed:
[CHORE] Apply <string-array>

<string-array> 적용

KWY0218 committed:
[REFACTOR] Remove button event in binding adapter

바인딩 어뎁터에 있던 버튼 이벤트 fragment에서 실행
fragment-ktx 이용해서 fragment replace 적용

KWY0218 committed:
[CHORE] Edit xml convention

카멜케이스에서 스네이크 케이스로 수정

KWY0218 committed:
[REFACTOR] Change from glide to coil

glide 화질 안 좋아서 coil로 변경

KWY0218 committed:
[REFACTOR] Remove init adapter in bindingadapter

bindingAdapter에서 했던 recyclerview adapter 연결 fragment에서 연결 하도록 변경

KWY0218 committed:
[CHORE] Change style to textappearance

style에서 textappearance로 변경

KWY0218 committed:
[CHORE] Change name

스네이크 네이밍으로 수정

KWY0218 committed:
[REFACTOR] Change LiveData to StateFlow

LiveData에서 StateFlow로 변경
util에 있던 Event 제거
bindingAdapter 통해서 에러 메시지 toast 출력
메모리 leak 때문에 activityViewModels 에서 viewmodels로 변경

KWY0218 committed:
[CHORE] 파일 정리

ui 단에 fragment가 너무 많아서
화면 기준으로 pakage 나눔

KWY0218 committed:
[CHORE] Edit result
